### PR TITLE
Apply some fusion-related hlints

### PR DESCRIPTION
**Pull request by jrp2014**
This set of fusing-related hlint tweaks passes the Travis test. They make things easier to read IMHO, if they don’t improve performance.

### CODE REVIEW

File: src/full/Agda/Auto/Convert.hs [hunk 1]
```diff
@@ -137,7 +137,7 @@ tomy imi icns typs = do
        let Datatype [con] [] = cdcont cc
        lift $ liftIO $ modifyIORef con (\cdef -> cdef {cdtype = contyp'})
 
-       projfcns <- mapM (\name -> getConst False name TMAll) (map I.unDom fields)
+       projfcns <- mapM ((\name -> getConst False name TMAll) . I.unDom) fields
```
Comment (andreasabel) [inline src/full/Agda/Auto/Convert.hs:140]: Wouldn't
```haskell
projfcns <- mapM (\ dom -> getConst False (I.unDom dom) TMAll) fields
```
be more readable?

File: src/full/Agda/Interaction/Highlighting/Generate.hs [hunk 2]
```diff
@@ -510,7 +510,7 @@ nameKinds hlLevel decl = do
     _      -> return HMap.empty
       -- Traverses the syntax tree and constructs a map from qualified
       -- names to name kinds. TODO: Handle open public.
-  let syntax = foldr ($) HMap.empty $ map declToKind $ universeBi decl
+  let syntax = foldr (($) . declToKind) HMap.empty (universeBi decl)
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/Interaction/Highlighting/Generate.hs:513]: Listen to hint-man!

File: src/full/Agda/Interaction/Highlighting/Generate.hs [hunk 3]
```diff
@@ -443,7 +443,7 @@ generateTokenInfoFromSource
      -- disk.
   -> TCM CompressedFile
 generateTokenInfoFromSource file input =
-  runPM $ tokenHighlighting <$> fst <$> Pa.parseFile Pa.tokensParser file input
+  runPM (tokenHighlighting . fst <$> Pa.parseFile Pa.tokensParser file input)
```
Comment (andreasabel) [inline src/full/Agda/Interaction/Highlighting/Generate.hs:446]: Avoid parentheses:
```haskell
runPM $ do tokenHighlighting . fst <$> Pa.parseFile Pa.tokensParser file input
```

File: src/full/Agda/Interaction/Highlighting/LaTeX.hs [hunk 4]
```diff
@@ -693,14 +693,12 @@ toLaTeX cc path source hi
 
   . map (second (
       -- Split tokens at newlines
-      concatMap stringLiteral
+      concatMap (stringLiteral
```
Comment (andreasabel) [inline src/full/Agda/Interaction/Highlighting/LaTeX.hs:696]: The structure of function compositions
```haskell
  f
  . g
  . h
```
is destroyed by this fusion.  You now how long ranging parenthesized expressions which are hard to read; the formatting also does not make sense anymore.

File: src/full/Agda/Interaction/Highlighting/Precise.hs [hunk 5]
```diff
@@ -301,7 +301,7 @@ compressedFileInvariant :: CompressedFile -> Bool
 compressedFileInvariant (CompressedFile []) = True
 compressedFileInvariant (CompressedFile f)  =
   all rangeInvariant rs &&
-  all (not . empty) rs &&
+  not (any empty rs) &&
```
Comment (andreasabel) [inline src/full/Agda/Interaction/Highlighting/Precise.hs:304]: Breaks the conjunctive pattern `all ... && all ... && and` (these are all conjunctions).

File: src/full/Agda/Syntax/Abstract/Name.hs [hunk 6]
```diff
@@ -197,7 +197,7 @@ showQNameId q = show ns ++ "@" ++ show m
 --   fallback when looking up the right concrete name in the scope fails.
 qnameToConcrete :: QName -> C.QName
 qnameToConcrete (QName m x) =
-  foldr C.Qual (C.QName $ nameConcrete x) $ map nameConcrete $ mnameToList m
+   foldr (C.Qual . nameConcrete) (C.QName $ nameConcrete x) (mnameToList m)
```
Comment (andreasabel) [inline src/full/Agda/Syntax/Abstract/Name.hs:200]: Worsens readability.

File: src/full/Agda/Syntax/Concrete/Operators.hs [hunk 7]
```diff
@@ -348,11 +348,10 @@ buildParsers r flat kind exprNames = do
     let g = Data.Function.fix $ \p -> Parsers
               { pTop    = memoise TopK $
                           Fold.asum $
-                            foldr ($) (pApp p)
-                              (map (\(l, ns) higher ->
-                                       mkP (Right l) parseSections
-                                           (pTop p) ns higher True)
-                                   relatedOperators) :
+                          foldr (($) . (\ (l, ns) higher
+                                        -> mkP (Right l) parseSections (pTop p) ns higher True))
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/Syntax/Concrete/Operators.hs:352]: Listen to hint-man.

File: src/full/Agda/TypeChecking/Coverage.hs [hunk 8]
```diff
@@ -782,7 +782,7 @@ createMissingHCompClause f n x old_sc (SClause tel ps _sigma' cps (Just t)) = se
                                                <#> (ilam "o" $ \ _ -> unEl <$> ty i) <@> u0 <@> u1
           alpha <- do
             vars <- mapM (open . applySubst hcompS . fst) alphab
-            return $ foldr imax (pure iz) $ map (\ v -> v `imax` ineg v) vars
+            return $ foldr (imax . (\ v -> v `imax` ineg v)) (pure iz) vars
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Coverage.hs:785]: Harder to read.

File: src/full/Agda/TypeChecking/Coverage.hs [hunk 9]
```diff
@@ -794,7 +794,7 @@ createMissingHCompClause f n x old_sc (SClause tel ps _sigma' cps (Just t)) = se
                                                             (ilam "o" $ \ _ -> apply_delta_fill i $ side1 <@> hfill lvl htype phi u u0 i))
              let recurse []           i = __IMPOSSIBLE__
                  recurse [(psi,u)]    i = u i
-                 recurse ((psi,u):xs) i = pOr_ty i psi (foldr imax (pure iz) (map fst xs)) (u i) (recurse xs i)
+                 recurse ((psi,u):xs) i = pOr_ty i psi ( foldr (imax . fst) (pure iz) xs) (u i) (recurse xs i)
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Coverage.hs:797]: I find this harder to read.

File: src/full/Agda/TypeChecking/Level/Solve.hs [hunk 10]
```diff
@@ -55,8 +55,7 @@ defaultLevelsToZero xs = loop =<< openLevelMetas (map MetaId $ IntSet.elems xs)
          | otherwise   -> return ()
 
     openLevelMetas :: [MetaId] -> m [MetaId]
-    openLevelMetas xs = return xs
-      >>= filterM (\m -> isNothing <$> isInteractionMeta m)
+    openLevelMetas xs =  filterM (\ m -> isNothing <$> isInteractionMeta m) xs
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/TypeChecking/Level/Solve.hs:58]: Here, we have our own idiom:
```haskell
filterM (isNothing <.> isInteractionMeta)
```
(See Agda.Utils.Functor.)

File: src/full/Agda/TypeChecking/Monad/Builtin.hs [hunk 11]
```diff
@@ -476,9 +476,9 @@ intervalUnview' :: HasBuiltins m => m (IntervalView -> Term)
 intervalUnview' = do
   iz <- fromMaybe __IMPOSSIBLE__ <$> getBuiltin' builtinIZero -- should it be a type error instead?
   io <- fromMaybe __IMPOSSIBLE__ <$> getBuiltin' builtinIOne
-  imin <- (`Def` []) <$> fromMaybe __IMPOSSIBLE__ <$> getPrimitiveName' "primIMin"
-  imax <- (`Def` []) <$> fromMaybe __IMPOSSIBLE__ <$> getPrimitiveName' "primIMax"
-  ineg <- (`Def` []) <$> fromMaybe __IMPOSSIBLE__ <$> getPrimitiveName' "primINeg"
+  imin <- ((`Def` []) . fromMaybe __IMPOSSIBLE__) <$> getPrimitiveName' "primIMin"
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/TypeChecking/Monad/Builtin.hs:479]: Listen to hint-man!

File: src/full/Agda/TypeChecking/Monad/Builtin.hs [hunk 12]
```diff
@@ -476,9 +476,9 @@ intervalUnview' :: HasBuiltins m => m (IntervalView -> Term)
...
+  imax <- ((`Def` []) . fromMaybe __IMPOSSIBLE__) <$> getPrimitiveName' "primIMax"
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/TypeChecking/Monad/Builtin.hs:480]: Again, spurious parentheses.

File: src/full/Agda/TypeChecking/Monad/Builtin.hs [hunk 13]
```diff
@@ -476,9 +476,9 @@ intervalUnview' :: HasBuiltins m => m (IntervalView -> Term)
...
+  ineg <- ((`Def` []) . fromMaybe __IMPOSSIBLE__) <$> getPrimitiveName' "primINeg"
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/full/Agda/TypeChecking/Monad/Builtin.hs:481]: Ditto.

File: src/full/Agda/TypeChecking/Monad/Constraints.hs [hunk 14]
```diff
@@ -39,7 +39,7 @@ solvingProblems pids m = verboseBracket "tc.constr.solve" 50 ("working on proble
 isProblemSolved :: (MonadTCEnv m, ReadTCState m) => ProblemId -> m Bool
 isProblemSolved pid =
   and2M (not . Set.member pid <$> asksTC envActiveProblems)
-        (all (not . Set.member pid . constraintProblems) <$> getAllConstraints)
+        ((not . any (Set.member pid . constraintProblems)) <$> getAllConstraints)
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Monad/Constraints.hs:42]: Symmetry broken: original code shows more clearly that this is a big conjunction.  What should be rather done here is to use `Set.notMember`.

File: src/full/Agda/TypeChecking/Monad/State.hs [hunk 15]
```diff
@@ -276,7 +276,7 @@ addRewriteRulesFor f rews matchables =
 
 setMatchableSymbols :: QName -> [QName] -> Signature -> Signature
 setMatchableSymbols f matchables =
-  foldr (.) id $ map (\g -> updateDefinition g $ setMatchable) matchables
+  foldr ((.) . (\ g -> updateDefinition g $ setMatchable)) id matchables
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Monad/State.hs:279]: Keep original code which is idiomatic (`foldr (.) id`) and easier to read.

File: src/full/Agda/TypeChecking/Primitive/Cubical.hs [hunk 16]
```diff
@@ -1259,7 +1259,7 @@ primTransHComp cmd ts nelims = do
             (flags,t_alphas) <- fmap unzip . forM as $ \ (bs,ts) -> do
                  let u' = listS bs' `applySubst` u
                      bs' = (Map.toAscList $ Map.map boolToI bs)
-                 let weaken = foldr composeS idS $ map (\ j -> liftS j (raiseS 1)) $ map fst bs'
+                 let weaken = foldr ((composeS . (\ j -> liftS j (raiseS 1))) . fst) idS bs'
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Primitive/Cubical.hs:1262]: -1

File: src/full/Agda/TypeChecking/Primitive/Cubical.hs [hunk 17]
```diff
@@ -1295,7 +1295,7 @@ primTransHComp cmd ts nelims = do
           combine l ty d [] = d
           combine l ty d [(psi,u)] = u
           combine l ty d ((psi,u):xs)
-            = pure tPOr <#> l <@> psi <@> (foldr imax iz (map fst xs))
+            = pure tPOr <#> l <@> psi <@> (foldr (imax . fst) iz xs)
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Primitive/Cubical.hs:1298]: -1

File: src/full/Agda/TypeChecking/Primitive/Cubical.hs [hunk 18]
```diff
@@ -1321,7 +1321,7 @@ primTransHComp cmd ts nelims = do
               let
                 phis :: [Term]
                 phis = for bools $ \ m ->
-                            foldr iMin iO $ map (\(i,b) -> if b then var i else iNeg (var i)) $ Map.toList m
+                            foldr (iMin . (\(i,b) -> if b then var i else iNeg (var i))) iO (Map.toList m)
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Primitive/Cubical.hs:1324]: -1

File: src/full/Agda/TypeChecking/Records.hs [hunk 19]
```diff
@@ -650,7 +650,7 @@ etaExpandAtRecordType t u = do
 {-# SPECIALIZE etaContractRecord :: QName -> ConHead -> ConInfo -> Args -> TCM Term #-}
 {-# SPECIALIZE etaContractRecord :: QName -> ConHead -> ConInfo -> Args -> ReduceM Term #-}
 etaContractRecord :: HasConstInfo m => QName -> ConHead -> ConInfo -> Args -> m Term
-etaContractRecord r c ci args = if all (not . usableModality) args then fallBack else do
+etaContractRecord r c ci args = if not (any usableModality args) then fallBack else do
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Records.hs:653]: -1

File: src/full/Agda/TypeChecking/Rules/Term.hs [hunk 20]
```diff
@@ -1544,7 +1544,7 @@ inferExprForWith e = do
 ---------------------------------------------------------------------------
 
 checkLetBindings :: [A.LetBinding] -> TCM a -> TCM a
-checkLetBindings = foldr (.) id . map checkLetBinding
+checkLetBindings =  foldr ((.) . checkLetBinding) id
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Rules/Term.hs:1547]: -1

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for this experiment, @jrp2014 !

I looked through the suggested changes and it seems like map-map fusion in general produces readable results, but not foldr-map fusion, and neither "not-all to any-not".
Thus, I suppose the good hints out of your selection would be:
- "functor law"
- "Fuse mapMaybe/map"
- "Fuse mapM/map"
- "Fuse mapM_/map"
- "Monad law, left identity"
We could try going with these.

File: src/full/Agda/Auto/Convert.hs (hunk 1 above)
Comment (WolframKahl) [reply src/full/Agda/Auto/Convert.hs:140]: How about:
```haskell
projfcns <- mapM (flip (getConst False) TMAll . I.unDom) fields
```
Or, equivalently:
```haskell
projfcns <- mapM ((getConst False `flip` TMAll) . I.unDom) fields
```
(Such infix ``` `flip` ``` can be quite useful, but probably takes getting used to...)

Or even:
```haskell
projfcns <- mapM (($ TMAll) . getConst False . I.unDom) fields
```
:wink:

Comment (jrp2014) [reply src/full/Agda/Auto/Convert.hs:140]: I don't have strong views on this but, as I've said elsewhere, it is more important that the same pattern is used throughout the codebase. I'm sure that it is possible to get hlint to rewrite in any of the suggested ways.  That said, flip is not my favourite function, but can see that it's useful if you prefer a completely pointfree approach.

**Comment by jrp2014** [comment]
Superseded by #4479

**Pull request closed without merging**